### ISSUE TITLE
Support nil cost components

### ISSUE THREAD

**Issue by sinabakh** [infracost/infracost#92]
There are cases where combining usage data with actual resources creates conditions that a cost component makes no sense. For instance, a provisioned DynamoDB shall not have `read/write request units`.
Here is a simple code for the `write request units` cost component:
```go
func NewDynamoDBTable(d *schema.ResourceData, u *schema.ResourceData) *schema.Resource {
  costComponents := make([]*schema.CostComponent, 0)
  costComponents = append(costComponents, wruCostComponent(d))
}

func wruCostComponent(d *schema.ResourceData) *schema.CostComponent {
  costComponent := &schema.CostComponent{}
  if billingMode == "PROVISIONED" {
    log.Warnf("Skipping write request units usage data for %s. Provisioned mode does not support write request units.", d.Address)
    return costComponent
  }
  ...
  return costComponent
}

```
This way, the core produces a warning for the empty cost component and the output shows an empty line.

The other way around is to return `nil` and check the result of the function and skip it if it's nil but it makes the code filled with the same conditions.

I propose we add support for `nil` cost components in the core.

**Comment by sinabakh** [comment]
I'll work on its PR so we can discuss it better.

**Comment by aliscott** [comment]
Is there an advantage of this over wrapping the append in a condition like this?
```go
costComponents := make([]*schema.CostComponent, 0)
if billingMode != "PROVISIONED" {
  costComponents = append(costComponents, wruCostComponent(d))
}
```

**Comment by sinabakh** [comment]
For this specific condition, it makes sense to wrap it inside an if clause. But there may be some more complicated conditions where you have to either:
* Check the return value of the cost component function
* Write some logic outside the cost component function

I can't think of such conditions for now, but I think it would be great not to have any specific logic outside of the cost component function. So in your example my idea is sth like:
```
cc = wruCostComponent(d)
if cc != nil {
  costComponents = append(costComponents, cc)
}
```

To sum it up, there are two design decisions to be made:
1. Add/Do not add nil/empty Cost Components.
2. If we are going to not to add them, what's the best way to handle it to make the code as clean as possible.

**Comment by aliscott** [comment]
I'm leaning towards not adding empty/nil cost components to favour explicitness.

**Comment by sinabakh** [comment]
Seems ok to me. Thanks @aliscott
